### DISCUSSION TITLE
Publishing a release from GitHub actions does not publish to GitHub marketplace

### DISCUSSION THREAD

**Discussion by untitaker** [community/community#7941] [Actions] [answered]
**Summary**

Publishing a release for a repo with an action.yml from GitHub actions will not update the version tag on the GitHub marketplace. The workaround I found is to manually edit the just-published release, uncheck and check the "Publish to Marketplace" option and save the release.

I imagine one could automate the workaround using undocumented APIs, but I have not investigated that option.

I have brought this up with support at the beginning of this year and they both confirmed my workaround as recommended for the time being, and relayed my feedback.

The original message is attached below, all I wrote there is still relevant. Support number found in the footer of emails is: ERX4KY-2VQV

----

> Hi. I wanted to publish https://github.com/untitaker/hyperlink to the GitHub marketplace. For that I edited the last release to check off the "publish to marketplace" option. So far so good. 0.1.14 is now published.

> Then I tried to publish another release by pushing a tag. This triggers .github/workflows/release.yml to create a release. When I look into this release, the option to publish to marketplace is checked: https://github.com/untitaker/hyperlink/releases/edit/0.1.15

> However, the marketplace still shows 0.1.14: https://github.com/marketplace/actions/hyperlink-link-checker

> I guess I will be able to solve this by unchecking and checking "publish to marketplace". However, ideally I would like to automate the entire release creation process.

> Which brings me to the actual question: Is there a way I can publish to GitHub marketplace using actions/create-release action (or other action)?

**Comment by MylesBorins** [comment]
hey @untitaker. I'm trying to hunt down the right folks from the marketplace team to take a look at this.

**Comment by azooinmyluggage** [comment]
@saquibkhan can you please take a look?

**Comment by saquibkhan** [comment]
Looping @priyakewlani18 and @ushakiran-r for initial triage

**Comment by priyakewlani18** [comment]
Hey @MylesBorins , 

We tried the same scenario and it's working as expected.
Steps which we followed:
1. Release an action using `action.yml` to the marketplace with tag `1.1` - observed `1.1` tag on the marketplace.
2. Again create a new release with the new tag `1.2` and marketplace check is by default enabled. We didn't uncheck and check the "Publish to the marketplace" option - observed a new tag `1.2` on the marketplace.

Please check the screen recording for the same. Let me know if we are missing something.

cc: @saquibkhan , @ushakiran-r 
https://user-images.githubusercontent.com/65468026/146415898-136efa7f-cba6-47cf-82fe-017930441b21.mov

**Comment by untitaker** [comment]
@priyakewlani18 Sorry, I guess that wasn't entirely clear from the initial post, but your scenario is actually a different one. I use a GitHub actions workflow to publish my marketplace action from CI. There is no Draft Release button I ever click on.

https://github.com/untitaker/hyperlink/blob/main/action.yml <- this is the action I want to publish

https://github.com/untitaker/hyperlink/blob/main/.github/workflows/release.yml <- this is the github workflow that uses another action to publish that first action

That second GitHub Actions workflow creates a release of the repo [using an official action](https://github.com/untitaker/hyperlink/blob/main/.github/workflows/release.yml#L15). This however creates a release that is not automatically published to marketplace.

To be clear, why do I go through such contortions? It's because releasing my project includes:

1. building a binary in CI
2. publishing it to github releases
3. publishing the new version of actions.yml from the repo to github marketplace

I want to achieve this by pushing a single tag. and it almost works, except the action's version does not get updated on the github marketplace page.

**Reply by rafaelschlatter** [reply]
I am still experiencing this problem. Workaround suggested by @untitaker works, but I prefer no manual action. Anyone have figured out how to solve this?

**Comment by SimonTheLeg** [comment]
Is there any update on this? I am also running into this issue and manually updating each release feels very painful to me

From my understanding, all the aforementioned actions make use of GH's public releases API (https://docs.github.com/en/rest/releases/releases#create-a-release). From the public documentation, it does not seem like there is an endpoint to control marketplace releases directly.

However the GH frontend does seem to use the same API. So I did a Google Chrome HAR Dump to find out what the GH frontend does. From the dump it seems that it sends a `published_on_marketplace` parameter in the body (twice even).

Could it be that it is just missing in the public documentation, but the functionality is already in place?

Here's an excerpt of the HAR Dump 

```json
{
  "request": {
    "method": "POST",
    "url": "https://github.com/SimonTheLeg/semver-tag-from-pr-action/releases/tag/v1.3.2",
    "postData": {
      "mimeType": "application/x-www-form-urlencoded",
      "text": "<redacted>",
      "params": [
        {
          "name": "_method",
          "value": "put"
        },
        {
          "name": "authenticity_token",
          "value": "<redacted>"
        },
        {
          "name": "release%5Bid%5D",
          "value": "72253844"
        },
        {
          "name": "release%5Bdraft%5D",
          "value": "0"
        },
        {
          "name": "release%5Brepository_action_release_attributes%5D%5Bpublished_on_marketplace%5D",
          "value": "0"
        },
        {
          "name": "release%5Brepository_action_release_attributes%5D%5Bpublished_on_marketplace%5D",
          "value": "1"
        },
        {
          "name": "repository_action%5Bprimary_category_id%5D",
          "value": "2"
        },
        {
          "name": "repository_action%5Bsecondary_category_id%5D",
          "value": "17"
        },
        {
          "name": "release%5Btag_name%5D",
          "value": "v1.3.2"
        },
        {
          "name": "release%5Btarget_commitish%5D",
          "value": "main"
        },
        {
          "name": "release%5Bname%5D",
          "value": "v1.3.2"
        },
        {
          "name": "generated_notes_state",
          "value": "initial"
        },
        {
          "name": "release%5Bprevious_tag_name%5D",
          "value": ""
        },
        {
          "name": "release%5Bbody%5D",
          "value": "<shortened>"
        },
        {
          "name": "release%5Brelease_assets_attributes%5D%5B%5D%5Bname%5D",
          "value": ""
        },
        {
          "name": "release%5Brelease_assets_attributes%5D%5B%5D%5Bid%5D",
          "value": ""
        },
        {
          "name": "release%5Brelease_assets_attributes%5D%5B%5D%5B_destroy%5D",
          "value": ""
        },
        {
          "name": "release%5Bprerelease%5D",
          "value": "0"
        }
      ]
    }
  }
}

```

**Reply by katexochen** [reply]
Did you manage to use this field via API or gh CLI?

**Reply by SimonTheLeg** [reply]
Not really unfortunately. I was hoping GH would give a reply what the correct way is, so I could create a PR for the release action I am using.

**Comment by rickstaa** [comment]
Pasting this here as a reference:

> I've asked internally and the answer is that publishing to Marketplace cannot be done via API at the moment because it requires a 2FA confirmation step, which is right now performed exclusively through one's web browser. The confirmation step is required for increased Marketplace security.
>
> In the future it could be possible that the 2FA step could be performed from the terminal as well, but I don't know when that will be possible at earliest. In the interest of cleaning our issue backlog, I'm going to close this for now.
>
> For Releases published via `gh release create` (or generally through the GitHub API), the way to also publish them to Marketplace in the meantime is to edit the release and uncheck/check the "Marketplace" option.

_Originally posted by @mislav in https://github.com/cli/cli/issues/5193#issuecomment-1036278907_

**Comment by untitaker** [comment, marked as answer]
I believe this issue got fixed sometime in 2023. I just had a publish to marketplace that worked correctly, using cargo-dist in untitaker/hyperlink.

**Reply by simoneb** [reply]
I confirm that I'm seeing the same behavior, without changing anything in the release creation logic, that we're doing via the REST API. Now actions released via the API are automatically published to the marketplace, as far as I've been able to notice.

**Reply by SharkBaitDLS** [reply]
Doesn't appear to be working for on my side still, but we're using an app token to create our releases. It seems like since app's can't have 2FA, it requires a human to go manually edit and re-publish the release to get the 2FA flow for the publish to go through.
